community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/LuceneIndexWriter.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.SearcherFactory;
+import org.apache.lucene.search.SearcherManager;
+import org.apache.lucene.store.Directory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+public class LuceneIndexWriter implements Closeable
+{
+    // Lucene cannot allocate a full MAX_INT of documents, the deviation differs from JVM to JVM, but according to
+    // their source in future versions, the deviation can never be bigger than 128.
+    private static final long MAX_DOC_LIMIT = Integer.MAX_VALUE - 128;
+
+    private static final String KEY_STATUS = "status";
+    private static final String ONLINE = "online";
+    private static final Map<String,String> ONLINE_STATUS_USER_DATA = Collections.singletonMap( KEY_STATUS, ONLINE );
+
+    final IndexWriter writer;
+
+    LuceneIndexWriter( Directory dir, IndexWriterConfig conf ) throws IOException
+    {
+        this.writer = new IndexWriter( dir, conf );
+    }
+
+    static boolean isOnline( Directory directory ) throws IOException
+    {
+        if ( !IndexReader.indexExists( directory ) )
+        {
+            return false;
+        }
+
+        try ( IndexReader reader = IndexReader.open( directory ) )
+        {
+            Map<String,String> userData = reader.getIndexCommit().getUserData();
+            return ONLINE.equals( userData.get( KEY_STATUS ) );
+        }
+    }
+
+    public void addDocument( Document document ) throws IOException
+    {
+        writer.addDocument( document );
+    }
+
+    public void updateDocument( Term term, Document document ) throws IOException
+    {
+        writer.updateDocument( term, document );
+    }
+
+    public void deleteDocuments( Term term ) throws IOException
+    {
+        writer.deleteDocuments( term );
+    }
+
+    public void deleteDocuments( Query query ) throws IOException
+    {
+        writer.deleteDocuments( query );
+    }
+
+    public SearcherManager createSearcherManager() throws IOException
+    {
+        return new SearcherManager( writer, true, new SearcherFactory() );
+    }
+
+    public void optimize() throws IOException
+    {
+        writer.optimize();
+    }
+
+    public void commit() throws IOException
+    {
+        writer.commit();
+    }
+
+    IndexDeletionPolicy getIndexDeletionPolicy()
+    {
+        return writer.getConfig().getIndexDeletionPolicy();
+    }
+
+    void commitAsOnline() throws IOException
+    {
+        writer.commit( ONLINE_STATUS_USER_DATA );
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        writer.close( true );
+    }
+
+    void reserveDocumentAdditions( int additionsCount ) throws IOException
+    {
+        throw new UnsupportedOperationException( "Should not be called" );
+    }
+
+    void withdrawReservationOfDocumentAdditions( int additionsCount )
```

Comment:
I'm not liking "withdraw" here, how about "release"? seems more in line with the word "reservation".